### PR TITLE
SystemProperty Test Helpers #524

### PR DESCRIPTION
**Pull request by sksamuel**

### CODE REVIEW

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package io.kotlintest.assertions
+
+import java.util.*
+
+fun <T> withSystemProperty(key: String, value: String, thunk: () -> T): T {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt]: Perhaps allow value to be nullable? One might want to override a specific System property to null

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    return thunk()
+  } finally {
+    if (previous == null)
+      System.clearProperty(key)
+    else
+      System.setProperty(key, previous)
+  }
+}
+
+fun <T> withSystemProperties(props: List<Pair<String, String?>>, thunk: () -> T): T {
```
Comment (LeoColman) [inline kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt:17]: Also add an overload to `Map<String, String?>`?

**Comment by LeoColman** [comment]
Do you think it's worth to add a `SystemPropertyExtension` of some sort?

**Comment by LeoColman** [comment]
Missing documentation on how to use this

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt (hunk 1 above)
Comment (sksamuel) [reply kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt]: Done.

File: kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt (hunk 2 above)
Comment (sksamuel) [reply kotlintest-assertions/src/main/kotlin/io/kotlintest/assertions/sysprops.kt:17]: Done.

**Comment by sksamuel** [comment]
> Do you think it's worth to add a `SystemPropertyExtension` of some sort?

Added system extension.

Documentation will come as a separate PR.

File: kotlintest-core/build.gradle [hunk 3]
```diff
@@ -1,4 +1,5 @@
 dependencies {
+    compile project(':kotlintest-assertions')
```
Comment (LeoColman) [inline kotlintest-core/build.gradle:2]: We had a discussion about this a while back. Do you think it's ok for core to depend  on assertions now?

**Comment by sksamuel** [comment]
Yes. The reason they're split is for people who want to use our assertions
but with junit (or spek even!).
But if you're using our test framework, then you're not going to mind one
more small dep.

**Review by LeoColman — APPROVED** [review]

**Pull request merged**